internal/topology/check/compatibility.go

```diff
@@ -54,3 +62,261 @@ func ObjectsAreInTheSameNamespace(current, desired client.Object) error {
+
+	return allErrs
+}
+
+// MachineDeploymentClassesAreCompatible checks if each MachineDeploymentClass in the new ClusterClass is a compatible change from the previous ClusterClass.
+// It checks if:
+// 1) Any MachineDeploymentClass has been removed.
+// 2) If the MachineDeploymentClass.Template.Infrastructure reference has changed its Group or Kind.
+func MachineDeploymentClassesAreCompatible(old, in *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+
+	// Ensure no MachineDeployment class was removed.
+	classes := classNamesFromWorkerClass(in.Spec.Workers)
+	for _, oldClass := range old.Spec.Workers.MachineDeployments {
+		if !classes.Has(oldClass.Class) {
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("spec", "workers", "machineDeployments"),
+					in.Spec.Workers.MachineDeployments,
+					fmt.Sprintf("The %q MachineDeployment class can't be removed.", oldClass.Class),
+				),
+			)
+		}
+	}
+
+	// Ensure previous MachineDeployment class was modified in a compatible way.
+	for i, class := range in.Spec.Workers.MachineDeployments {
+		for _, oldClass := range old.Spec.Workers.MachineDeployments {
+			if class.Class == oldClass.Class {
+				// NOTE: class.Template.Metadata and class.Template.Bootstrap are allowed to change;
+				// class.Template.Bootstrap are ensured syntactically correct by validateAllRefs.
+				// Validates class.Template.Infrastructure template changes in a compatible way
+				if err := TemplatesAreCompatible(class.Template.Infrastructure, oldClass.Template.Infrastructure); err != nil {
+					allErrs = append(allErrs,
+						field.Invalid(field.NewPath("spec", "workers", "machineDeployments").Index(i).Child("template", "infrastructure"),
+							class.Template.Infrastructure.Ref,
+							err.Error(),
+						),
+					)
+				}
+			}
+		}
+	}
+	return allErrs
+}
+
+// MachineDeploymentClassesAreUnique checks that no two MachineDeploymentClasses in a ClusterClass share a name.
+func MachineDeploymentClassesAreUnique(in *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+	classes := sets.NewString()
+	for i, class := range in.Spec.Workers.MachineDeployments {
+		if classes.Has(class.Class) {
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("spec", "workers", "machineDeployments").Index(i).Child("class"),
+					class.Class,
+					fmt.Sprintf("MachineDeployment class should be unique. MachineDeployment with class %q is defined more than once.", class.Class),
+				),
+			)
+		}
+		classes.Insert(class.Class)
+	}
+	return allErrs
+}
+
+// MachineDeploymentClassesAreUniqueAndDefinedInClusterClass checks that each MachineDeploymentClass is defined in the passed ClusterClass.
+func MachineDeploymentClassesAreUniqueAndDefinedInClusterClass(in *clusterv1.Cluster, class *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+	if in.Spec.Topology.Workers == nil {
+		return nil
+	}
```

Comment:
```suggestion
	}

```

Maybe add a similar guard clause for len(MachineDeployment) == 0

Reply:
Isn't this alright because the loop will just escape and we'll return errors anyway?

Reply:
Yup, it's just way easier to see that way :)